email_strides/app.py

```diff
@@ -0,0 +1,183 @@
+
+    This lambda will run at the beginning of the month, looking at the
+    previous month and comparing change to the month before that.
+
+    The Start date is inclusive, and the End date is exclusive
+    """
+    target_period = {}
+    compare_period = {}
+
+    # Special-case the two cases where we cross year boundaries
+    if today.month == 1:
+        # in Jan, look at Dec and Nov of last year
+        target_period["Start"] = f"{today.year - 1}-12-01"
+        target_period["End"] = f"{today.year}-01-01"
+
+        compare_period["Start"] = f"{today.year - 1}-11-01"
+        compare_period["End"] = f"{today.year - 1}-12-01"
+
+    elif today.month == 2:
+        # in Feb, look at Jan of this year and Dec of last year
+        target_period["Start"] = f"{today.year}-01-01"
+        target_period["End"] = f"{today.year}-02-01"
+
+        compare_period["Start"] = f"{today.year - 1}-12-01"
+        compare_period["End"] = f"{today.year}-01-01"
+
+    else:
+        # no year boundary, look at the previous two months
+        target_period["Start"] = f"{today.year}-{(today.month - 1):02}-01"
+        target_period["End"] = f"{today.year}-{today.month:02}-01"
+
+        compare_period["Start"] = f"{today.year}-{(today.month - 2):02}-01"
+        compare_period["End"] = f"{today.year}-{(today.month - 1):02}-01"
+
+    LOG.info(f"Target month: {target_period}")
+    LOG.info(f"Compare month: {compare_period}")
+
+    return target_period, compare_period
+
+
+def parse_results_by_time(results_by_time, compare=None):
+    """
+    Transform results returned from Cost Explorer into a useful data structure,
+    and optionally calculating change from a previous result.
+    """
+    data = {}
+
+    minimum = float(os.environ["MINIMUM"])
+
+    for result in results_by_time:
+        for group in result["Groups"]:
+            amount = float(group["Metrics"][ce.cost_metric]["Amount"])
+            if amount < minimum:
+                continue
+
+            if len(group["Keys"]) != 1:
+                LOG.error(f"Unexpected grouping: {group['Keys']}")
+                continue
+
+            key = group["Keys"][0]
+            data[key] = {"total": amount}
+
+            if compare and key in compare:
+                _total = data[key]["total"]
+                _compare = compare[key]["total"]
+
+                # changes from zero are special cases
+                if _compare == 0:
+                    if _total == 0:
+                        # both are zero, no change
+                        pct = 0
+                    else:
+                        # up from zero, 100% change
+                        pct = 1
+                else:
+                    # calculate percent change
+                    pct = (_total / _compare) - 1
+            else:
+                pct = 1.0
+
+            data[key]["change"] = pct
+
+    return data
+
+
+def get_service_costs(target_period, compare_period):
```

Comment:
by "different information" I mean that these CE queries are different than the other reports, and parsing the response requires knowing the length of the `GroupBy` value.